controllers/sms-gateway.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * This module implements GET and POST to support medic-gateway's API
+ * @see https://github.com/medic/medic-gateway
+ */
+
+var _ = require('underscore'),
+    db = require('../db'),
+    http = require('http'),
+    querystring = require('querystring'),
+    utils = require('./utils'),
+    messageUtils = require('./messages');
+
+require('lie/polyfill');
+
+var UNUSED = null; // this var is declared in various functions in the messages controller but never actually used
+
+function readBody(stream) {
+  var body = '';
+  return new Promise(function(resolve, reject) {
+    stream.on('data', function(data) {
+      body += data.toString();
+    });
+    stream.on('end', function() {
+      resolve(body);
+    });
+    stream.on('error', reject);
+  });
+}
+
+function readResBody(res) {
+  return readBody(res)
+    .then(function(body) {
+      if (res.statusCode >= 400) {
+        throw new Error('Bad status received: ' + res.statusCode + ': ' + body);
+      }
+      return body;
+    });
+}
+
+function saveToDb(gatewayRequest, wtMessage) {
+  var messageBody = querystring.stringify({
+    from: wtMessage.from,
+    message: wtMessage.content,
+    'medic-gateway_id': wtMessage.id,
+  });
+
+  new Promise(function(resolve, reject) {
+    var req = http.request(
+      {
+        hostname: 'localhost',
+        port: 5984,
+        path: '/medic/_design/medic/_rewrite/add',
+        method: 'POST',
+        headers: {
+          'Content-Length': messageBody.length,
+          'Content-Type': 'application/x-www-form-urlencoded',
+          'Authorization': gatewayRequest.headers.authorization,
+        },
+      },
+      function(res) {
+        readResBody(res)
+          .then(JSON.parse)
+          .then(function(response) {
+            return resolve();
+          })
+          .catch(reject);
+      });
+
+    req.on('error', reject);
+
+    req.write(messageBody);
+    req.end();
+  })
+  .catch(function(err) {
+    console.log('sms-gateway', 'saveToDb', 'error saving', wtMessage, err);
+  });
+}
+
+function getWebappState(update) {
+  switch(update.status) {
+    case 'SENT':
+      return 'sent';
+    case 'DELIVERED':
+      return 'delivered';
+    case 'FAILED':
+      return 'failed';
+  }
+}
+
+function updateStateFor(gatewayRequest, update) {
+  var newState = getWebappState(update);
+  if (!newState) {
+    return Promise.reject(new Error('Could not work out new state for update: ' + JSON.stringify(update)));
+  }
+
+  return updateState(
+      gatewayRequest,
+      gatewayRequest.headers['user-agent'],
+      update.id,
+      newState);
+}
+
+function updateState(gatewayRequest, userAgent, messageId, newState) {
+  var updateBody = {
+    state: newState,
+    details: {
+      useragent: userAgent,
```

Comment:
Yeah i will ditch it for now - don't know of any current projects which have multiple gateway phones.